File: python/tradingeconomics/worldBank.py
import json 
import urllib 
import pandas as pd
import sys
from datetime import *
from . import functions as fn
from . import glob
import ssl


PY3 = sys.version_info[0] == 3

if PY3: # Python 3+
    from urllib.request import urlopen
    from urllib.parse import quote
else: # Python 2.X
    from urllib import urlopen
    from urllib import quote


class ParametersError(ValueError):
    pass

class CredentialsError(ValueError):
    pass

class LoginError(AttributeError):
    pass

class WebRequestError(ValueError):
    pass


def checkSeriesCode(linkAPI, series_code):
    linkAPI = 'https://api.tradingeconomics.com/worldBank/indicator'     
    if series_code == None:
        linkAPI += '?c=' + glob.apikey + '&s=' + quote(str(series_code)) 
    else:
        linkAPI += '?c=' + glob.apikey + '&s=' + quote("".join(series_code)) 
       
    return linkAPI  
        
def checkPageNumber(linkAPI, page_number):
    if page_number != None:
        linkAPI +=  '/{0}'.format(page_number) 
    return linkAPI

def checkCountry(linkAPI, country):
    linkAPI = 'https://api.tradingeconomics.com/worldBank/country/'       
    if type(country) is str:
        linkAPI += quote(str(country), safe='')
    else:
        linkAPI += quote(",".join(country), safe='')
        
    return linkAPI

def checkIndicator(linkAPI, indicator): 
    linkAPI = 'https://api.tradingeconomics.com/worldBank/indicator/'     
    if type(indicator) is str:
        linkAPI +=  quote(str(indicator), safe='')
    else:
        linkAPI += quote(",".join(str(indicator), safe=''))
    
    return linkAPI
 
    

def getWBCategories(category = None, page_number = None, output_type = None):
    """
    Return a list of all categories, categories by page number.
    =================================================================================

    Parameters:
    -----------
    categories:list.
             List of strings for all categories and list of categories by page number.
             All categories, for example:
                category = None
             Several categories or one category, for example:
                category = ['education', 'agriculture']
                category = 'education'
             categories by page, for example:
                category = 'education', page_number = 3
                category = ['education', 'agriculture'], page_number = 5   
    output_type: string.
             'dict'(default) for dictionary format output, 'df' for data frame,
             'raw' for list of dictionaries directly from the web. 

    Notes
    -----
    All parameters are optional. Without parameters a list of all categories will be provided. 

    Example
    -------
    getWBCategories(category = None, output_type = None)

    getWBCategories(category = ['education', 'agriculture'], output_type = None)
    """          
    url = ''
    try:
        _create_unverified_https_context = ssl._create_unverified_context
    except AttributeError:
        pass
    else:
        ssl._create_default_https_context = _create_unverified_https_context
    if category:
        url = 'https://api.tradingeconomics.com/worldBank/category/' + quote(str(category), safe='')
    else: 
        url = 'https://api.tradingeconomics.com/worldBank/categories'
    
    if category == None:
        linkAPI = 'https://api.tradingeconomics.com/worldBank/categories'
    else:
        linkAPI = 'https://api.tradingeconomics.com/worldBank/category/' + quote(str(category), safe='')
    
    if page_number != None:
        linkAPI = checkPageNumber(linkAPI, page_number)
     
    try:
        linkAPI += '?c=' + glob.apikey
    except AttributeError:
        raise LoginError('You need to do login before making any request')
    
    try:
        # print(linkAPI)
        return fn.dataRequest(api_request=linkAPI, output_type=output_type)
    except Exception as e:
        print(e) 
        

def getWBIndicator(series_code = None, url = None, output_type = None):
    """
    Detailed information about specific indicator for all countries using a series 
    code or url.
    =================================================================================

    Parameters:
    -----------
    series_code:list.
             List of strings of indicators by series code.
             sring of indicator by country using it's url.
             Specific indicator and country by using series code, for example:
                series_code = 'usa.fr.inr.rinr'(symbol used by TE for a country)
    Url:string.            
             Specific indicator and country by using url, for example:
                url = '/united-states/real-interest-rate-percent-wb-data.html'             
    output_type: string.
             'dict'(default) for dictionary format output, 'df' for data frame,
             'raw' for list of dictionaries directly from the web. 

    Notes
    -----
    A series code or url is required. 

    Example
    -------

    getWBIndicator(series_code = 'usa.fr.inr.rinr', url = None, output_type = None)

    getWBIndicator(series_code = None, url = '/united-states/real-interest-rate-percent-wb-data.html', output_type = None)
    """          
    try:
        _create_unverified_https_context = ssl._create_unverified_context
    except AttributeError:
        pass
    else:
        ssl._create_default_https_context = _create_unverified_https_context
   
    linkAPI = 'https://api.tradingeconomics.com/worldBank/indicator/'  
    if series_code == None and url == None:
        return "Series code or url is required!"  

    if series_code != None:
        linkAPI = 'https://api.tradingeconomics.com/worldBank/indicator' + '?c=' + glob.apikey + '&s=' + quote(str(series_code), safe='')            
    elif url != None:
        linkAPI = 'https://api.tradingeconomics.com/worldBank/indicator' + '?c=' + glob.apikey + '&url=' + quote(str(url), safe='')
         
   
    try:
        # print(linkAPI)
        return fn.dataRequest(api_request=linkAPI, output_type=output_type)
    except Exception as e:
        print(e) 

  
def getWBCountry(country = None, page_number = None, output_type = None):
    """
    List of indicators available for a specific country (with pagination).
    =================================================================================

    Parameters:
    -----------
    country:list.
             List of strings of indicators by country.
             country, for example:
                country = 'portugal'
    page_number:            
                country = 'portugal', page_number = 3               
    output_type: string.
             'dict'(default) for dictionary format output, 'df' for data frame,
             'raw' for list of dictionaries directly from the web. 

    Notes
    -----
    A country is required to get a list.

    Example
    -------
    getWBCountry(country = 'portugal', output_type = None) # page_number is no longer needed!
    """ 
    linkAPI = 'https://api.tradingeconomics.com/worldBank/country/'          
    try:
        _create_unverified_https_context = ssl._create_unverified_context
    except AttributeError:
        pass
    else:
        ssl._create_default_https_context = _create_unverified_https_context
  
    if country == None:
        return "A country is required!" 
    else:
        linkAPI = checkCountry(linkAPI, country)

    if page_number != None:    
        linkAPI = checkPageNumber(linkAPI, page_number)
    
    try:
        linkAPI += '?c=' + glob.apikey
    except AttributeError:
        raise LoginError('You need to do login before making any request')
    
    try:
        # print(linkAPI)
        return fn.dataRequest(api_request=linkAPI, output_type=output_type)
    except Exception as e:
        print(e) 


def getWBHistorical(series_code = None, output_type = None):
    """
    Historical data for a specific indicator.
    =================================================================================

    Parameters:
    -----------
    series_code:list.
             List of historical data by country.
             for example:
                series_code = None
                series_code = 'usa.fr.inr.rinr'            
    output_type: string.
             'dict'(default) for dictionary format output, 'df' for data frame,
             'raw' for list of dictionaries directly from the web. 

    Notes
    -----
    A series code is required.

    Example
    -------
    
    getWBHistorical(series_code = 'usa.fr.inr.rinr', output_type = None)
    """          
    try:
        _create_unverified_https_context = ssl._create_unverified_context
    except AttributeError:
        pass
    else:
        ssl._create_default_https_context = _create_unverified_https_context
    
    linkAPI = 'https://api.tradingeconomics.com/worldBank/historical' 
    
    if series_code == None:
        return "A series code is required!"
    else:
        linkAPI += '?c=' + glob.apikey + '&s=' + quote(str(series_code))   
    
    try:
        # print(linkAPI)
        return fn.dataRequest(api_request=linkAPI, output_type=output_type)
    except Exception as e:
        print(e) 
  
    

File: R/tradingeconomics/R/worldBank.R
source("R/functions.R")


#'Get World Bank values from Trading Economics API
#'@export getWorldBankCategories
#'
#' @param category string.

#' @param outType string.
#''df' for data frame,
#''raw'(default) for list of unparsed data.
#'
#'@return Returns a list or data from World Bank.
#'@section Notes:
#'Without credentials only sample information will be provided. Without a category, a list of all categories will be provided.
#'@seealso \code{\link{getCalendarData}}, \code{\link{getForecastData}}, \code{\link{getHistoricalData}} and \code{\link{getIndicatorData}}
#'@examples
#'\dontrun{ getWorldBankCategories(outType = 'df')
#'getWorldBankCategories(category = 'education')
#'getWorldBankCategories(category = 'education')
#'}


getWorldBankCategories <- function( category = NULL, outType = NULL){

  base <- "https://api.tradingeconomics.com/worldBank"
  url <- paste(base)
  df_final = data.frame()
  apikey_local <- .GlobalEnv$apiKey
  if(!is.null(category)){
    url<- paste(base, 'category', paste(category, collapse = ','), sep = '/')
  }else{
    url <- paste(base, 'categories', sep = '/')
  }

  url <- paste(url,  '?c=', apikey_local, sep ='')

  url <- URLencode(url)
  request <- GET(url)
  checkRequestStatus(http_status(request)$message)

  webResults <- do.call(rbind.data.frame, checkForNull(content(request)))
  df_final = rbind(df_final, webResults)
  Sys.sleep(0.5)


  if (is.null(outType)| identical(outType, 'lst')){
    df_final <- split(df_final , f =  paste(df_final$category))
  } else if (identical(outType, 'df')){
    if (length(df_final) == 0){
      print ("No data provided for selected parameters")
    }
  } else {
    stop('output_type options : df for data frame, lst(default) for list by category')
  }

  return(df_final)

}


#'Get World Bank values from Trading Economics API
#'@export getWorldBankIndicators
#'
#' 
#'@param category string.
#'@param identifier string.
#'@param URL string.
#'@param series_code string.
#'@param outType string.
#''df' for data frame,
#''raw'(default) for list of unparsed data.
#'
#'@return Returns a list or data from World Bank.
#'@section Notes:
#'Detailed information about specific indicator for all countries using a series code , for a specific country with pagination, for a specific
#'indicator and country by using series_code or url.
#'@seealso \code{\link{getCalendarData}}, \code{\link{getForecastData}}, \code{\link{getHistoricalData}} and \code{\link{getIndicatorData}}
#'@examples
#'\dontrun{ getWorldBankIndicators(series_code = 'fr.inr.rinr', outType = 'df')
#'getWorldBankIndicators(identifier = 'portugal',category = 'country', outType = 'df')
#'getWorldBankIndicators(series_code = 'usa.fr.inr.rinr')
#'getWorldBankIndicators(URL = '/united-states/real-interest-rate-percent-wb-data.html')
#'}


getWorldBankIndicators <- function(category = NULL, identifier = NULL, series_code = NULL,  URL = NULL, outType = NULL){
  base <- "https://api.tradingeconomics.com/worldBank"
  url <- paste(base)
  df_final = data.frame()


  if(!is.null(identifier)) #if indicator is something
  {
    if(is.null(category) || category == "country") {
      url <- paste(base,"country", paste(identifier, collapse = ','), sep = '/')
    } else if (category == "symbol"){
      url<- paste(base,"symbol", paste(identifier, collapse = ','), sep = '/')
    } else {
      return("Category not recognized. It can be null, country or symbol.")
    }
  }
  apikey_local <- .GlobalEnv$apiKey
  url <- paste(url,  '?c=', apikey_local, sep ='')

  if(!is.null(series_code)){
    url <- paste(base,"indicator", paste(identifier, collapse = ','), sep = '/')
    url <- paste(url,  '?c=', apikey_local, sep ='')
    url <- paste(url, paste(series_code, collapse = ','), sep = '&s=')
  }

  if(!is.null(URL)){
    url <- paste(base,"indicator", collapse = NULL, sep = '/')
    url <- paste(url,  '?c=', apikey_local, sep ='')
    url <- paste(url, paste(URL, collapse = ','), sep = '&url=')
    print(url)
  }



  url <- URLencode(url)
  request <- GET(url)
  checkRequestStatus(http_status(request)$message)

  webResults <- do.call(rbind.data.frame, checkForNull(content(request)))
  df_final = rbind(df_final, webResults)
  Sys.sleep(0.5)


  if (is.null(outType)| identical(outType, 'lst')){
    df_final <- split(df_final , f =  paste(df_final$category, df_final$URL))
  } else if (identical(outType, 'df')){
    if (length(df_final) == 0){
      print ("No data provided for selected parameters")
    }
  } else {
    stop('output_type options : df for data frame, lst(default) for list by category')
  }

  return(df_final)

}

#'Get World Bank values from Trading Economics API
#'@export getWorldBankHistorical
#'
#' @param series_code string.
#'
#' @param outType string.
#''df' for data frame,
#''raw'(default) for list of unparsed data.
#'
#'@return Returns a list or data from World Bank.
#'@section Notes:
#'Series_code must be provided.
#'@seealso \code{\link{getCalendarData}}, \code{\link{getForecastData}}, \code{\link{getHistoricalData}} and \code{\link{getIndicatorData}}
#'@examples
#'\dontrun{ getWorldBankHistorical('usa.fr.inr.rinr')
#'
#'}
#'

getWorldBankHistorical <- function(series_code = NULL, outType = NULL){
  base <- "https://api.tradingeconomics.com/worldBank/historical"
  apikey_local <- .GlobalEnv$apiKey
  url <- paste(base, '?c=', apikey_local, sep = '')
  df_final = data.frame()

  if(!is.null(series_code)){
    url <- paste(url, paste(series_code, collapse = ','), sep = '&s=')
  }

  url <- URLencode(url)
  request <- GET(url)


  checkRequestStatus(http_status(request)$message)

  webResults <- do.call(rbind.data.frame, checkForNull(content(request)))

  df_final = rbind(df_final, webResults)
    Sys.sleep(0.5)


  if (is.null(outType)| identical(outType, 'lst')){
    df_final <- split(df_final , f = paste(df_final$symbol))
  } else if (identical(outType, 'df')){
    df_final = df_final
  } else {
    stop('output_type options : df for data frame, lst(default) for list by country ')
  }

  return(df_final)
}

